### PR TITLE
librustc: hint close matches on accessing nonexisting fields

### PR DESCRIPTION
**Pull request by oli-obk**
rebase and fix of #19267

### CODE REVIEW

**Comment by oli-obk** [comment]
something must have changed in rust, since adding those help messages make the error fatal, and there are unit tests checking if the error is not fatal.

**Comment by oli-obk** [comment]
the error message uses `fcx.type_error_message` while the hints use `tcx.sess.span_help`
i did not find a help equivalent to the type_error_message function

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3572,11 +3573,44 @@ fn check_expr_with_unifier<'a, 'tcx, F>(fcx: &FnCtxt<'a, 'tcx>,
                             actual)
                 },
                 expr_t, None);
+            if let Some(t) = ty::ty_to_def_id(expr_t) {
+                suggest_field_names(t, field, tcx, vec![]);
+            }
```
Comment (alexcrichton) [inline src/librustc_typeck/check/mod.rs]: Could you add a test for this error message as well?

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3572,11 +3573,44 @@ fn check_expr_with_unifier<'a, 'tcx, F>(fcx: &FnCtxt<'a, 'tcx>,
...
+    // displays hints about the closest matches in field names
+    fn suggest_field_names<'tcx>(id : DefId,
+                                 field : &ast::SpannedIdent,
+                                 tcx : &ty::ctxt<'tcx>,
+                                 skip : Vec<&str>) {
+        let ident = token::get_ident(field.node);
+        let name = ident.get();
+        // only find fits with at least one matching letter
+        let mut best_dist = name.len();
+        let mut best = vec![];
```
Comment (alexcrichton) [inline src/librustc_typeck/check/mod.rs]: I believe that the corresponding code in resolve only prints at most one suggestion, perhaps we could do that for now?

File: src/librustc_typeck/check/mod.rs (hunk 1 above)
Comment (oli-obk) [inline src/librustc_typeck/check/mod.rs]: sure, allthough the test is not useful since it's ignored:  #21195

**Comment by oli-obk** [comment]
rebased on master, added test and changed number of help messages to one. local make check-stage1 passes, make check is running right now

**Comment by oli-obk** [comment]
make check passed

**Comment by alexcrichton** [comment]
Thanks! Could you squash the commits together as well?

**Comment by oli-obk** [comment]
i just left them there for easier reading.
squashed and rebased

**Comment by oli-obk** [comment]
sorry about that. fixed, rebased and make check ran through

**Pull request merged**